### ISSUE TITLE
Models are not fully scikit-learn compatible

### ISSUE THREAD

**Issue by lfleck** [scikit-multiflow/scikit-multiflow#45]
In order to follow scikit-learn's philosophy, wouldn't it make sense to derive your base model https://github.com/scikit-multiflow/scikit-multiflow/blob/c16892a053c6f033c4408ab9e418607b8b764919/src/skmultiflow/core/base.py#L5 

from the sklearn base class for estimators http://scikit-learn.org/stable/modules/generated/sklearn.base.BaseEstimator.html ?

Since `get_params` and `set_params` are not implemented in scikit-multiflow, useful sklearn utils like http://scikit-learn.org/stable/modules/generated/sklearn.base.clone.html can not be applied to the scikit-multiflow models.

**Comment by jacobmontiel** [comment]
The main concern here is to bring things that are not compatible with `scikit-multiflow`.

The safest option would be to bring those elements that are relevant such as the above mentioned `get_params`, `set_params`

Comments @jmread , @lfleck ?

**Comment by remiadon** [comment]
Hi there, 
I also have difficulties understanding the API of scikit-multiflow, 
`StreamModel` inherits `BaseObject`
both of these classes implement scikit-learn basic functions (fit, partial_fit, etc..) as `abstractmethod`s, raising NotImplementedError by default

I think it would be simpler, and also closer to sklearn's philosophy, if models in skmultiflow just inherit the `BaseModel` class from sklearn.

For the moment I don't see why a StreamModel class should be implemented, or I am missing a point maybe ?

I thing a good example of this is the `PerceptronMask ` class, in 
[/src/skmultiflow/neural_networks/perceptron.py](/src/skmultiflow/neural_networks/perceptron.py)

> Because scikit-multiflow's framework require a few interfaces, not present 
    int scikit-learn, this mask allows the first to use classifiers native to 
    the latter

From what I understand this class is just a wrapper on top of sklearn's Perceptron
Ideally skmultiflow should not redefine this class to make it compatible with its own API, but rather provide an API natively compatible with sklearn

**Comment by jacobmontiel** [comment]
Finally got some time to properly look into this. Here is some brainstorming:

* `clone` seems a robust way to generate new estimators with the default parameters. This is currently achieved in `skmultiflow` via the `reset` method. However the benefit from `clone` is that it is inherited by all extended classes whereas `reset` is not. In this case we also need `get_params` and `set_params`
* The `__repr__` method is a similar case, functionality is currently provided by `get_info` at the extended classes level.
* Regarding the `mixins`, it is interesting to see that `fit`, `partial_fit`, `predict` and `predict_proba` are not defined. In this case I prefer to keep the definitions as in `skmultiflow`. However, as new functionality is added, the benefits of task-specific mixins (classification, regression, etc.) are considerable. For example, regression models have in the `partial_fit` signature the `classes` parameter which is not relevant for this type of learning. 
* There is some other functionality which is not (at this moment) applicable to our framework. In this case I prefer to add it when (and if) it is necessary rather than bring it directly.

> I think it would be simpler, and also closer to sklearn's philosophy...

This is a very important feature and something we definetly want to ensure. 

> ...if models in skmultiflow just inherit the BaseModel class from sklearn.

Here is were things get a little complicated. `skmultiflow` is not intended as an extension of sklearn but as a complement. Although we want to provide compatibility with (incremental) methods from sklearn, our main objective is to extend the options for the user. In this sense, `skmultiflow` specializes on the setting for incremental methods which is not the case in `sklearn`.

The perceptron wrapper was one of the first attemps in terms of framework compatibility. As you can see, the wrapper just provides the above mentioned methods (`get_info` and `reset`). However, it is important to mention that even if we add this functionality it does not guarantee 100% compatibility. For example, some incremental methods in `sklearn` require to call `fit` before attempting a `partial_fit`. Another difference is that calling `predict` on a `sklearn` model before fitting (or partially fitting) it will raise and error. 

Comments?

**Comment by remiadon** [comment]
#
> clone seems a robust way to generate new estimators [...] we also need get_params and set_params

Why not including reset (and/or) clone inside a BaseEstimator class proper to skmultiflow ?
```python
class BaseStreamEstimator(BaseEstimator): 
    def reset(self, *args, **kwargs):
         return self.__init__(*args, **kwargs)
    def clone(self):
        pass  # TODO ?
```

This way, if I go back to my use case (sklearn`s perceptron), this becomes
```python
from sklearn.linear_model.perceptron import Perceptron
class StreamPerceptron(Perceptron, BaseStreamEstimator)
    pass

sp = StreamPerceptron()
sp.reset()  # OK
```

#
> Regarding the mixins, it is interesting to see that fit, partial_fit, predict and predict_proba are not defined [...]

True, sklearn does not include `fit`, `partial_fit` nor `predict` via inheritance from the mixin classes. On the other hand skearn does provide separate Mixins for classification and regression (see [here](https://github.com/scikit-learn/scikit-learn/blob/7b136e9/sklearn/base.py#L131)). Maybe skmultiflow could do the same ? Could this help with something like regression models having `classes` as parameter ?

**Comment by jacobmontiel** [comment]
I have created a branch to work on this change: https://github.com/scikit-multiflow/scikit-multiflow/tree/Improved_base_class

There is now an initial version including mixins for classification and regression. 

Thanks for the suggestion @remiadon , I like the concept of `BaseStreamEstimator` because it also helps differentiate additional functionality not available in `sklearn`

I have replaced the `StreamModel` for `BaseStreamEstimator` in `NaiveBayes` as a starting point. I will continue updating the rest of the methods.

**Comment by remiadon** [comment]
@jacobmontiel don't hesitate if you need help on this one

**Comment by jacobmontiel** [comment]
Progress on this issue is steady and will be complete very soon. Anybody feel free to check the branch in case I am missing something.

So far I have not found a pythonic way to implement a generic `reset` method. Furthermore, I have noticed that not all the tests cover the multiple implementations of this method.

**Comment by remiadon** [comment]
@jacobmontiel Could you create a PR so we can add comments ?

**Comment by jacobmontiel** [comment]
Thanks for the suggestion @remiadon 
 
I have created the PR

**Comment by jacobmontiel** [comment]
Here is an update on the new/improved base classes as they stand today in the branch:

* `BaseEstimator` is the equivalent of `sklearn.base.BaseEstimator` without extras unrelated to the stream/skmultiflow setting
* `BaseSKMObject` is an extension of `BaseEstimator` that implements additional functionality specific to `skmultiflow`.
* `StreamModel` is replaced by a combination of `BaseSKMObject` and task-specific mixins: `ClassifierMixin`, `RegressorMixin`, `MetaEstimatorMixin`, `MultiOutputMixin`, ....
* Legacy classes `BaseObject` and `StreamModel` have been deleted
* `Stream`, `StreamEvaluator` and `BaseDriftDetector` are now extensions of `BaseSKMObject`

Following are diagrams of the new class-hierarchies.

### Base classes with methods
![base_wmethods](https://user-images.githubusercontent.com/17923265/57482707-9f579c00-72a5-11e9-8800-fcf6486f96b0.png)

### `Stream`, `StreamEvaluator` and `BaseDriftDetector` hierarchy 
![base_stream_and_evaluator_drift_detector](https://user-images.githubusercontent.com/17923265/57482903-1b51e400-72a6-11e9-8f2e-b766c697f476.png)

### Example of hierarchy for a classifier (Naive Bayes)
![naive_bayes_diagram](https://user-images.githubusercontent.com/17923265/57482708-9f579c00-72a5-11e9-9431-85caec82b9ce.png)

Comments @remiadon @jmread @abifet ?
